// @ccint/sg_new/webpack.component.js
const path = require('path')
const merge = require('webpack-merge')
const common = require('./webpack.common.js')

module.exports = merge(common, {
  entry: {
    'page': './src/components/page.js',
    'dialog': './src/components/dialog.js',
    'datepicker': './src/components/datepicker.js',
    'fileuploader': './src/components/fileuploader.js',
    'switch': './src/components/switch.js',
    'router': './src/components/router.js',
    'toast': './src/components/toast.js',
    'placeholder': './src/components/placeholder.js',
    'autocomplete': './src/components/autocomplete.js',
    'select': './src/components/select.js',
    'uploader_pc': './src/components/uploader_pc.js',
    'listview': './src/components/listview.js',
    'actionsheet': './src/components/actionsheet.js',
    'rollview': './src/components/rollview.js',
    'datepickermo': './src/components/datepickermo.js'
  },
  output: {
    filename: '[name].min.js',
    libraryTarget: 'umd',
    libraryExport: 'default',
    path: path.resolve(__dirname, 'dist')
  },
  externals: {
    jquery: {
      commonjs: 'jquery',
      commonjs2: 'jquery',
      amd: 'jquery',
      root: '$'
    },
    '../sg.js': {
      commonjs: '@ccint/sg_new',
      commonjs2: '@ccint/sg_new',
      amd: '@ccint/sg_new',
      root: 'sg'
    },
    'art-template/lib/template-web.js': {
      commonjs: 'art-template/lib/template-web.js',
      commonjs2: 'art-template/lib/template-web.js',
      amd: 'art-template/lib/template-web.js',
      root: 'template'
    },
    'exif-js': {
      commonjs: 'exif-js',
      commonjs2: 'exif-js',
      amd: 'exif-js',
      root: 'EXIF'
    }
  }
})
